src/examples/_03_solabi.rs

```diff
@@ -0,0 +1,111 @@
+use ethrpc::{eth, types::*};
+use futures::future::join_all;
+use solabi::{decode::Decode, encode::Encode, selector, Address, FunctionEncoder};
+use std::collections::HashMap;
+
+const NAME: FunctionEncoder<(), (String,)> = FunctionEncoder::new(selector!("name()"));
+const SYMBOL: FunctionEncoder<(), (String,)> = FunctionEncoder::new(selector!("symbol()"));
+const TOKEN_URI: FunctionEncoder<(U256,), (String,)> =
+    FunctionEncoder::new(selector!("tokenURI(uint256)"));
+
+fn name_call(address: Address) -> TransactionCall {
+    TransactionCall {
+        to: Some(address),
+        input: Some(NAME.encode_params(&())),
+        ..Default::default()
+    }
+}
+
+/// Using solabi docs example:
+fn symbol_call(address: Address) -> TransactionCall {
+    TransactionCall {
+        to: Some(address),
+        input: Some(SYMBOL.encode_params(&())),
+        ..Default::default()
+    }
+}
+
+fn uri_call(token: NftId) -> TransactionCall {
+    TransactionCall {
+        to: Some(token.address),
+        input: Some(TOKEN_URI.encode_params(&(token.id,))),
+        ..Default::default()
+    }
+}
+
+fn decode_function_result_string<T>(
+    res: &Result<Vec<u8>, ethrpc::http::Error>,
+    encoder: FunctionEncoder<T, (String,)>,
+) -> Option<String>
+where
+    T: Encode + Decode,
+{
+    match res {
+        Ok(bytes) => Some(encoder.decode_returns(bytes).expect("always blue").0),
+        Err(err) => {
+            tracing::warn!("got {:?}", err.to_string());
+            None
+        }
+    }
+}
+
+pub async fn get_name_and_symbol(
+    addresses: Vec<Address>,
+) -> HashMap<Address, (Option<String>, Option<String>)> {
+    let client = ethrpc::http::Client::from_env().buffered(Default::default());
+    tracing::info!("Preparing {} GetName Requests", addresses.len());
+    let name_futures = addresses
+        .iter()
+        .cloned()
+        .map(|addr| client.call(eth::Call, (name_call(addr), BlockId::default())));
+
+    let symbol_futures = addresses
+        .iter()
+        .cloned()
+        .map(|addr| client.call(eth::Call, (symbol_call(addr), BlockId::default())));
+
+    let names = join_all(name_futures).await;
+    let symbols = join_all(symbol_futures).await;
```

Comment:
You are first awaiting all names, and once that is finished you will start awaiting all symbols. A future does not start executing until it is being polled for the first time, and the creation of this iterator does not do any polling on the contained futures. Parallel awaiting is trivial though:

```rs
use futures::future::{join, join_all};

let (names, symbols) = join(
  join_all(name_futures),
  join_all(symbol_futures),
).await;
```

Reply:
Thanks! This is great!